api/run: does not work for a run without a commit

As of #1134, GET `/api/run/{run_id}` won't work for a run when the run's commit is None. Example: this test I just wrote:
https://github.com/conbench/conbench/blob/ee16a7946f88812a996dc9dd53c08e8673b8445e/conbench/tests/api/test_runs.py#L92-L111
fails with
```
conbench/api/runs.py:73: in get
    return self.serializer.one._dump(run, get_baseline_runs=True)
conbench/entities/run.py:417: in _dump
    commit = CommitSerializer().one.dump(run.commit)
conbench/entities/_entity.py:219: in dump
    return self._dump(data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <conbench.entities.commit._Serializer object at 0xffff885f7610>, commit = None

    def _dump(self, commit):
        url = None
>       if commit.repository and commit.sha:
E       AttributeError: 'NoneType' object has no attribute 'repository'
```

--- a/conbench/entities/run.py
+++ b/conbench/entities/run.py
@@ -414,9 +414,17 @@ def commit_fetch_info_and_create_in_db_if_not_exists(
 
 class _Serializer(EntitySerializer):
     def _dump(self, run: Run, get_baseline_runs: bool = False):
-        commit = CommitSerializer().one.dump(run.commit)
+        if run.commit:
+            commit = CommitSerializer().one.dump(run.commit)
+            commit.pop("links", None)
+            commit_link = f.url_for(
+                "api.commit", commit_id=commit["id"], _external=True
+            )
+        else:
+            commit = None
+            commit_link = None
+
         hardware = HardwareSerializer().one.dump(run.hardware)
-        commit.pop("links", None)
         hardware.pop("links", None)
         result = {
             "id": run.id,
@@ -439,14 +447,13 @@ class _Serializer(EntitySerializer):
             "links": {
                 "list": f.url_for("api.runs", _external=True),
                 "self": f.url_for("api.run", run_id=run.id, _external=True),
-                "commit": f.url_for(
-                    "api.commit", commit_id=commit["id"], _external=True
-                ),
                 "hardware": f.url_for(
                     "api.hardware", hardware_id=hardware["id"], _external=True
                 ),
             },
         }
+        if commit_link:
+            result["links"]["commit"] = commit_link
         if get_baseline_runs:
             result["candidate_baseline_runs"] = run.get_candidate_baseline_runs()
         return result

--- a/conbench/tests/api/test_runs.py
+++ b/conbench/tests/api/test_runs.py
@@ -89,6 +89,27 @@ class TestRunGet(_asserts.GetEnforcer):
             ),
         )
 
+    def test_get_run_without_commit(self, client):
+        self.authenticate(client)
+        result = _fixtures.benchmark_result(no_github=True)
+        run = result.run
+        response = client.get(f"/api/runs/{run.id}/")
+        self.assert_200_ok(
+            response,
+            _expected_entity(
+                run,
+                candidate_baseline_runs={
+                    "fork_point": DEFAULT_BRANCH_PLACEHOLDER,
+                    "latest_default": DEFAULT_BRANCH_PLACEHOLDER,
+                    "parent": {
+                        "baseline_run_id": None,
+                        "commits_skipped": None,
+                        "error": "no matching baseline run was found",
+                    },
+                },
+            ),
+        )
+
     def test_get_run_should_not_prefer_test_runs_as_baseline(self, client):
         """Test runs shouldn't be preferred, but if they are the only runs that exist,
         we'll pick them up"""

--- a/conbench/tests/api/test_runs.py
+++ b/conbench/tests/api/test_runs.py
@@ -17,8 +17,8 @@ DEFAULT_BRANCH_PLACEHOLDER = {
 }
 
 
-def _expected_entity(run, candidate_baseline_runs=None):
-    parent = run.commit.get_parent_commit()
+def _expected_entity(run: Run, candidate_baseline_runs=None):
+    parent = run.commit.get_parent_commit() if run.commit else None
     has_errors = False
     entity = _api_run_entity(
         run.id,
@@ -94,21 +94,29 @@ class TestRunGet(_asserts.GetEnforcer):
         result = _fixtures.benchmark_result(no_github=True)
         run = result.run
         response = client.get(f"/api/runs/{run.id}/")
-        self.assert_200_ok(
-            response,
-            _expected_entity(
-                run,
-                candidate_baseline_runs={
-                    "fork_point": DEFAULT_BRANCH_PLACEHOLDER,
-                    "latest_default": DEFAULT_BRANCH_PLACEHOLDER,
-                    "parent": {
-                        "baseline_run_id": None,
-                        "commits_skipped": None,
-                        "error": "no matching baseline run was found",
-                    },
+        expected = _expected_entity(
+            run,
+            candidate_baseline_runs={
+                "fork_point": {
+                    "baseline_run_id": None,
+                    "commits_skipped": None,
+                    "error": "the contender run is not connected to the git graph",
                 },
-            ),
+                "latest_default": {
+                    "baseline_run_id": None,
+                    "commits_skipped": None,
+                    "error": "this baseline commit type does not exist for this run",
+                },
+                "parent": {
+                    "baseline_run_id": None,
+                    "commits_skipped": None,
+                    "error": "the contender run is not connected to the git graph",
+                },
+            },
         )
+        expected["commit"] = None
+        expected["links"].pop("commit")
+        self.assert_200_ok(response, expected)
 
     def test_get_run_should_not_prefer_test_runs_as_baseline(self, client):
         """Test runs shouldn't be preferred, but if they are the only runs that exist,

